syntax = "proto3";

package posts;

message Post {
  int32 id = 1;
  string title = 2;
  string description = 3;
  int32 creator_id = 4;
  string created_at = 5;
  string updated_at = 6;
  bool private = 7;
  repeated string tags = 8;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  int32 creator_id = 3;
  bool private = 4;
  repeated string tags = 5;
}

message CreatePostResponse {
  Post post = 1;
}

message UpdatePostRequest {
  int32 post_id = 1;
  string title = 2;
  string description = 3;
  bool private = 4;
  repeated string tags = 5;
  int32 creator_id = 6;
}

message UpdatePostResponse {
  Post post = 1;
}

message DeletePostRequest {
  int32 post_id = 1;
  int32 creator_id = 2;
}

message DeletePostResponse {
  bool success = 1;
}

message GetPostRequest {
  int32 post_id = 1;
}

message GetPostResponse {
  Post post = 1;
}

message ListPostsRequest {
  int32 page = 1;
  int32 count = 2;
}

message ListPostsResponse {
  repeated Post posts = 1;
}

service PostService {
  rpc CreatePost(CreatePostRequest) returns (CreatePostResponse);
  rpc UpdatePost(UpdatePostRequest) returns (UpdatePostResponse);
  rpc DeletePost(DeletePostRequest) returns (DeletePostResponse);
  rpc GetPost(GetPostRequest) returns (GetPostResponse);
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
}
